core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkedEncodedInputStream.java

```diff
@@ -0,0 +1,303 @@
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.auth.aws.internal.chunkedencoding.Chunk;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.Validate;
+
+
+/**
+ * An implementation of chunk-transfer encoding, but by wrapping an {@link InputStream}.
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ * <p>
+ * Per <a href="https://datatracker.ietf.org/doc/html/rfc7230#section-4.1">RFC-7230</a>, a chunk-transfer encoded message
+ * is defined as:
+ * <pre>
+ *     chunked-body   = *chunk
+ *                      last-chunk
+ *                      trailer-part
+ *                      CRLF
+ *     chunk          = chunk-size [ chunk-ext ] CRLF
+ *                      chunk-data CRLF
+ *     chunk-size     = 1*HEXDIG
+ *     last-chunk     = 1*("0") [ chunk-ext ] CRLF
+ *     chunk-data     = 1*OCTET ; a sequence of chunk-size octets
+ * </pre>
+ */
+@SdkProtectedApi
+public final class ChunkedEncodedInputStream extends InputStream {
+    private static final Logger LOG = Logger.loggerFor(ChunkedEncodedInputStream.class);
+    private static final byte[] CRLF = {'\r', '\n'};
+    private static final byte[] END = {};
+
+    private final InputStream inputStream;
+    private final int chunkSize;
+
+    private final ChunkHeader header;
+    private final List<ChunkExtension> extensions = new ArrayList<>();
+    private final List<Trailer> trailers = new ArrayList<>();
+
+    private Chunk currentChunk;
+    private boolean isFinished = false;
+
+    private ChunkedEncodedInputStream(BuilderImpl builder) {
+        this.inputStream = Validate.notNull(builder.inputStream, "Input-Stream cannot be null!");
+        this.chunkSize = Validate.isPositive(builder.chunkSize, "Chunk-size must be greater than 0!");
+        this.header = Validate.notNull(builder.header, "Header cannot be null!");
+        this.extensions.addAll(Validate.notNull(builder.extensions, "Extensions cannot be null!"));
+        this.trailers.addAll(Validate.notNull(builder.trailers, "Trailers cannot be null!"));
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (currentChunk == null || !currentChunk.hasRemaining() && !isFinished) {
+            currentChunk = getChunk(inputStream);
+        }
+
+        return currentChunk.stream().read();
+    }
+
+    /**
+     * Get an encoded chunk from the input-stream, or the final chunk if we've reached the end of the input-stream.
+     */
+    private Chunk getChunk(InputStream stream) throws IOException {
+        LOG.debug(() -> "Reading next chunk.");
+        if (currentChunk != null) {
+            currentChunk.close();
+        }
+        // we *have* to read from the backing stream in order to figure out if it's the end or not
```

Comment:
On the other hand, it's a big improvement in performance if we don't have to read the chunk data from the backing stream. It reduces the amount of copying of the data that we need to do.

Instead of reading the whole chunk to figure out if it's the end, can we wrap `inputStream` in a [`LookaheadInputStream`](https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/LookaheadInputStream.java) and peek() at the next `read()` result to see if it's `-1`? If it is, then do `getFinalChunk()`?

We can save this for a future pre-release task, but I'm being picky since this is a protected API.

Reply:
I don't quite understand. We have to read from the stream to use it when transforming the chunk (the whole chunk of data needs to be there, i.e. when chunk-signature is calculated). It's not clear to me (right now) what benefit a `LookaheadInputStream` would bring in this case?

Reply:
Maybe I'm not thinking it through enough and I need to see this being used in the real world to understand. Thanks for your patience.

My point was that we shouldn't need the whole chunk in memory to calculate a checksum that is needed in the final trailer. We could theoretically provide each byte to the checksum implementation as it is read out of the input stream.

With this CR, when someone reads data from the ChunkedEncodingInputStream, the data gets copied twice: Once on line 94 into an in-memory buffer specific to the chunked encoding input stream and again into whatever buffer the result of read() is written into.

If we defer that until read() and we use the target buffer that the caller provides (assuming we override the other overloads of read(), which we should for performance), we can just copy the data once: into the result buffer.

We don't necessarily need to implement that right now, since what we have works, but since it's a protected API we need for it to be possible. I don't think it is right now, since everywhere the chunk is provided it's already been read into memory so there's no chance we could fix that later.

Reply:
Discussed offline. We'll table this for now, add a TODO, as we will keep these classes internal so that the implementations can be changed around.